File: src/rootbox/pty.c
/* pty.c - PTY handling and I/O */
#include "rootbox.h"

void restore_tty(int fd, struct termios *saved) {
    if (isatty(fd)) {
        tcsetattr(fd, TCSANOW, saved);
    }
}

int setup_pty(int *master_fd, int *slave_fd) {
    char slave_name[PATH_MAX];
    struct termios tio;
    struct winsize ws;
    
    /* Get current terminal settings to copy them */
    int has_tty = isatty(STDIN_FILENO);
    if (has_tty) {
        if (tcgetattr(STDIN_FILENO, &tio) < 0) {
            fprintf(stderr, "warning: tcgetattr failed: %s\n", strerror(errno));
            has_tty = 0;
        }
        if (ioctl(STDIN_FILENO, TIOCGWINSZ, &ws) < 0) {
            fprintf(stderr, "warning: TIOCGWINSZ failed: %s\n", strerror(errno));
            ws.ws_row = 24;
            ws.ws_col = 80;
        }
    } else {
        /* Default terminal size */
        ws.ws_row = 24;
        ws.ws_col = 80;
    }
    
    /* Open PTY pair */
    if (openpty(master_fd, slave_fd, slave_name, has_tty ? &tio : NULL, &ws) < 0) {
        return -1;
    }
    
    return 0;
}

void setup_pty_slave(int slave_fd) {
    /* Become session leader */
    if (setsid() < 0) {
        fprintf(stderr, "warning: setsid failed: %s\r\n", strerror(errno));
    }
    
    /* Redirect stdio to the slave PTY */
    if (dup2(slave_fd, STDIN_FILENO) < 0) fatal("dup2 stdin failed");
    if (dup2(slave_fd, STDOUT_FILENO) < 0) fatal("dup2 stdout failed");
    if (dup2(slave_fd, STDERR_FILENO) < 0) fatal("dup2 stderr failed");
    
    if (slave_fd > STDERR_FILENO) {
        close(slave_fd);
    }
    
    /* Set controlling terminal */
    if (ioctl(STDIN_FILENO, TIOCSCTTY, 0) < 0) {
        fprintf(stderr, "warning: TIOCSCTTY failed: %s\r\n", strerror(errno));
    }
}

void io_loop(int master_fd, pid_t child_pid) {
    char buf[4096];
    fd_set readfds;
    int max_fd = (master_fd > STDIN_FILENO ? master_fd : STDIN_FILENO) + 1;
    
    while (1) {
        FD_ZERO(&readfds);
        FD_SET(STDIN_FILENO, &readfds);
        FD_SET(master_fd, &readfds);
        
        int ret = select(max_fd, &readfds, NULL, NULL, NULL);
        if (ret < 0) {
            if (errno == EINTR) continue;
            break;
        }
        
        /* Data from stdin -> pty master */
        if (FD_ISSET(STDIN_FILENO, &readfds)) {
            ssize_t n = read(STDIN_FILENO, buf, sizeof(buf));
            if (n <= 0) break;
            if (write(master_fd, buf, n) != n) break;
        }
        
        /* Data from pty master -> stdout */
        if (FD_ISSET(master_fd, &readfds)) {
            ssize_t n = read(master_fd, buf, sizeof(buf));
            if (n <= 0) break;
            if (write(STDOUT_FILENO, buf, n) != n) break;
        }
    }
}

File: src/rootbox-rs/src/pty.rs
use crate::config::Config;
use crate::error::{Result, RootboxError};
use nix::libc::{winsize, TIOCGWINSZ, TIOCSCTTY};
use nix::pty::openpty;
use nix::sys::termios;
use nix::sys::termios::{tcgetattr, tcsetattr, SetArg, Termios};
use nix::unistd::{dup2, setsid, Pid};
use std::io::{self};
use std::os::unix::io::{AsRawFd, BorrowedFd, RawFd};
use tracing::{debug, warn};

/// PTY manager for pseudo-terminal handling
pub struct PtyManager {
    config: Config,
    master_fd: Option<RawFd>,
    slave_fd: Option<RawFd>,
    original_termios: Option<Termios>,
}

impl PtyManager {
    pub fn new(config: Config) -> Self {
        Self {
            config,
            master_fd: None,
            slave_fd: None,
            original_termios: None,
        }
    }

    /// Setup PTY pair
    pub fn setup_pty(&mut self) -> Result<(RawFd, RawFd)> {
        if !self.config.features.pty_enabled {
            return Err(RootboxError::PtyError("PTY disabled in config".to_string()));
        }

        debug!("Setting up PTY");

        // Get terminal attributes if stdin is a TTY
        let stdin_fd = std::io::stdin().as_raw_fd();
        let term_attrs = if nix::unistd::isatty(stdin_fd).unwrap_or(false) {
            match tcgetattr(unsafe { BorrowedFd::borrow_raw(stdin_fd) }) {
                Ok(attrs) => {
                    self.original_termios = Some(attrs.clone());
                    Some(attrs)
                },
                Err(e) => {
                    warn!("Failed to get terminal attributes: {}", e);
                    None
                },
            }
        } else {
            None
        };

        // Get window size
        let winsize = self.get_window_size(stdin_fd);

        // Open PTY
        let pty_result = openpty(&winsize, term_attrs.as_ref())
            .map_err(|e| RootboxError::PtyError(format!("Failed to open PTY: {}", e)))?;

        let master_raw = pty_result.master.as_raw_fd();
        let slave_raw = pty_result.slave.as_raw_fd();

        self.master_fd = Some(master_raw);
        self.slave_fd = Some(slave_raw);

        debug!("PTY created: master={}, slave={}", master_raw, slave_raw);

        // Keep the OwnedFd alive by leaking them (they'll be closed manually)
        std::mem::forget(pty_result.master);
        std::mem::forget(pty_result.slave);

        Ok((master_raw, slave_raw))
    }

    /// Get current window size or use defaults
    fn get_window_size(
        &self,
        fd: RawFd,
    ) -> winsize {
        let mut ws: winsize = unsafe { std::mem::zeroed() };

        if nix::unistd::isatty(fd).unwrap_or(false) {
            unsafe {
                if libc::ioctl(fd, TIOCGWINSZ, &mut ws) == 0 {
                    return ws;
                }
            }
        }

        // Use default size from config
        ws.ws_row = self.config.pty.default_rows;
        ws.ws_col = self.config.pty.default_cols;
        ws
    }

    /// Setup slave PTY in child process
    pub fn setup_slave(
        &self,
        slave_fd: RawFd,
    ) -> Result<()> {
        debug!("Setting up slave PTY");

        // Create new session
        setsid().map_err(|e| {
            warn!("Failed to create new session: {}", e);
            RootboxError::PtyError(format!("Failed to create session: {}", e))
        })?;

        // Redirect stdio to slave PTY
        dup2(slave_fd, libc::STDIN_FILENO)
            .map_err(|e| RootboxError::PtyError(format!("Failed to dup2 stdin: {}", e)))?;

        dup2(slave_fd, libc::STDOUT_FILENO)
            .map_err(|e| RootboxError::PtyError(format!("Failed to dup2 stdout: {}", e)))?;

        dup2(slave_fd, libc::STDERR_FILENO)
            .map_err(|e| RootboxError::PtyError(format!("Failed to dup2 stderr: {}", e)))?;

        // Close slave fd if it's greater than stderr
        if slave_fd > libc::STDERR_FILENO {
            nix::unistd::close(slave_fd)
                .map_err(|e| RootboxError::PtyError(format!("Failed to close slave fd: {}", e)))?;
        }

        // Set controlling terminal
        unsafe {
            if libc::ioctl(libc::STDIN_FILENO, TIOCSCTTY, 0) != 0 {
                warn!(
                    "Failed to set controlling terminal: {}",
                    io::Error::last_os_error()
                );
            }
        }

        Ok(())
    }

    /// Set terminal to raw mode
    pub fn set_raw_mode(&self) -> Result<()> {
        let stdin_fd = std::io::stdin().as_raw_fd();

        if !nix::unistd::isatty(stdin_fd).unwrap_or(false) {
            return Ok(());
        }

        debug!("Setting terminal to raw mode");

        let mut termios = tcgetattr(unsafe { BorrowedFd::borrow_raw(stdin_fd) }).map_err(|e| {
            RootboxError::PtyError(format!("Failed to get terminal attributes: {}", e))
        })?;

        termios::cfmakeraw(&mut termios);

        tcsetattr(
            unsafe { BorrowedFd::borrow_raw(stdin_fd) },
            SetArg::TCSANOW,
            &termios,
        )
        .map_err(|e| RootboxError::PtyError(format!("Failed to set raw mode: {}", e)))?;

        Ok(())
    }

    /// Restore original terminal settings
    pub fn restore_terminal(&self) -> Result<()> {
        if let Some(ref termios) = self.original_termios {
            let stdin_fd = std::io::stdin().as_raw_fd();

            if nix::unistd::isatty(stdin_fd).unwrap_or(false) {
                debug!("Restoring terminal settings");

                tcsetattr(
                    unsafe { BorrowedFd::borrow_raw(stdin_fd) },
                    SetArg::TCSANOW,
                    termios,
                )
                .map_err(|e| {
                    warn!("Failed to restore terminal: {}", e);
                    RootboxError::PtyError(format!("Failed to restore terminal: {}", e))
                })?;
            }
        }

        Ok(())
    }

    /// Run I/O loop between master PTY and stdin/stdout (blocking version)
    pub fn io_loop_blocking(
        &self,
        master_fd: RawFd,
        _child_pid: Pid,
    ) -> Result<()> {
        use nix::sys::select::{select, FdSet};

        debug!("Starting I/O loop");

        let stdin_fd = std::io::stdin().as_raw_fd();
        let stdout_fd = std::io::stdout().as_raw_fd();
        let max_fd = std::cmp::max(stdin_fd, master_fd) + 1;

        let mut buf = [0u8; 4096];

        loop {
            let mut readfds = FdSet::new();
            // Safety: these file descriptors are valid
            unsafe {
                readfds.insert(BorrowedFd::borrow_raw(stdin_fd));
                readfds.insert(BorrowedFd::borrow_raw(master_fd));
            }

            // Use select to wait for data
            match select(max_fd, Some(&mut readfds), None, None, None) {
                Ok(_) => {},
                Err(nix::errno::Errno::EINTR) => continue,
                Err(e) => {
                    warn!("select failed: {}", e);
                    break;
                },
            }

            // Data from stdin -> pty master
            if unsafe { readfds.contains(BorrowedFd::borrow_raw(stdin_fd)) } {
                match nix::unistd::read(stdin_fd, &mut buf) {
                    Ok(0) => break, // EOF
                    Ok(n) => {
                        // Write to master using raw syscall
                        let written = unsafe {
                            libc::write(master_fd, buf[..n].as_ptr() as *const libc::c_void, n)
                        };
                        if written != n as isize {
                            warn!("Failed to write to master");
                            break;
                        }
                    },
                    Err(e) => {
                        warn!("Failed to read from stdin: {}", e);
                        break;
                    },
                }
            }

            // Data from pty master -> stdout
            if unsafe { readfds.contains(BorrowedFd::borrow_raw(master_fd)) } {
                match nix::unistd::read(master_fd, &mut buf) {
                    Ok(0) => break, // EOF
                    Ok(n) => {
                        // Write to stdout using raw syscall
                        let written = unsafe {
                            libc::write(stdout_fd, buf[..n].as_ptr() as *const libc::c_void, n)
                        };
                        if written != n as isize {
                            warn!("Failed to write to stdout");
                            break;
                        }
                    },
                    Err(nix::errno::Errno::EIO) => {
                        // EIO means the child process has exited - this is normal
                        break;
                    },
                    Err(e) => {
                        warn!("Failed to read from master: {}", e);
                        break;
                    },
                }
            }
        }

        debug!("I/O loop ended");
        Ok(())
    }

    /// Close master and slave file descriptors
    pub fn close(&mut self) -> Result<()> {
        if let Some(master) = self.master_fd.take() {
            let _ = nix::unistd::close(master);
        }

        if let Some(slave) = self.slave_fd.take() {
            let _ = nix::unistd::close(slave);
        }

        Ok(())
    }
}

impl Drop for PtyManager {
    fn drop(&mut self) {
        let _ = self.close();
        let _ = self.restore_terminal();
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_pty_manager_creation() {
        let config = Config::default();
        let manager = PtyManager::new(config);
        assert!(manager.master_fd.is_none());
    }
}
